test(gui-client): unit test for Linux IPC

ReactorScram opened the pull request:
(After GA)

This adds a unit test for the Unix domain sockets that I intend to use for process splitting on Linux.

The length-prefixed encoding and decoding are copied from `subzone`, but most of that code will not be re-used since it's Windows-specific and also specific to a Chromium-like process model, which won't work for Firezone.

ReactorScram committed:
test(gui-client): unit test for Linux IPC

ReactorScram committed:
not sure how to check UIDs yet

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into test/linux-ipc

ReactorScram committed:
fix UID check

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into test/linux-ipc

jamilbk commented on rust/client-tunnel/Cargo.toml:
```diff
@@ -6,6 +6,14 @@ edition = "2021"
+nix = { version =  "0.28.0", features = ["user"] }
+serde = { version = "1.0.197", features = ["derive"] }
+serde_json = "1.0.114"
+tokio = { version = "1.36.0", features = ["fs", "io-util", "macros", "net", "rt-multi-thread", "time"] }
```
It might be good to be more granular with these dependencies if they aren't required for other platforms, as our sleuthing session today discovered.

ReactorScram replied:
Yeah. If `version-check` has gone back to Linux or been removed completely, this doesn't have to compile on macOS at all, just Windows and Linux. So I can move `nix` into a Linux section for sure. I'll look into it.

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into test/linux-ipc

ReactorScram committed:
refactor yaml and make dependencies OS-specific

ReactorScram commented on .github/workflows/_rust.yml:
```diff
@@ -1,11 +1,15 @@
+---
 name: Rust
-on:
+"on":
```
This makes `yamllint` happier

ReactorScram commented on .github/workflows/_rust.yml:
```diff
@@ -19,11 +23,11 @@ jobs:
         include:
           - runs-on: ubuntu-22.04
-            packages: # Intentionally blank as a package catch-all linter
+            packages:  # Intentionally blank as a package catch-all linter
```
`yamllint` also said there should be two spaces before a comment here.

ReactorScram committed:
couldn't hurt to test this on macOS too

ReactorScram commented on .github/workflows/_rust.yml:
```diff
@@ -1,11 +1,15 @@
   run:
     working-directory: ./rust
 
+macos-packages: &macos-packages "-p connlib-client-apple -p connlib-client-shared -p firezone-tunnel -p snownet"
```
I added a couple packages that should be able to build and test on macOS, since they're built for prod anyway.

ReactorScram committed:
come on Github

ReactorScram committed:
fix

ReactorScram committed:
force Windows to use Bash

ReactorScram committed:
label non-cross-platform function

ReactorScram committed:
change up the scripts

- The `set -euo pipefail` in a script meant to be source was causing
  my interactive Bash to bail out. Only top-level scripts should set that
- Change the packages to "yes I'm word-splitting this on purpose" syntax

ReactorScram committed:
fix up the tests

ReactorScram committed:
name Github Action steps

ReactorScram committed:
Merge branch 'test/linux-ipc' of github.com:firezone/firezone into test/linux-ipc

ReactorScram committed:
fmt

ReactorScram commented on .github/workflows/_rust.yml:
```diff
@@ -19,23 +20,29 @@ jobs:
+            packages: macos-rust-packages.sh
           - runs-on: windows-2022
-            packages: -p connlib-client-shared -p firezone-gui-client -p snownet
+            packages: windows-rust-packages.sh
```
I refactored these because they were written in 3 places and I think already drifting out of sync

jamilbk replied:
Thanks for doing that!

What do you think of modifying the `.github/actions/setup-rust` action to just export the packages as a variable depending on runner.os instead?

That would keep CI configuration within `.github/` and out of `scripts/` and allow other Rust workflows to use the packages output if needed

jamilbk replied:
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action#creating-an-action-metadata-file

```
outputs:
  packages:
    description: Relevant packages for the current OS
    value: ${{
      (runner.os == 'Windows' && '...') ||
      (runner.os == 'Linux' && '...') ||
      (runner.os == 'macOS' && '...') }}
```

ReactorScram replied:
Yeah that might make it easier, and it would still show up in `git grep`. I'll try it.

jamilbk replied:
You might need single-quoted strings... I think GitHub might be quote-sensitive in some cases

ReactorScram commented on scripts/tests/lib.sh:
```diff
@@ -1,7 +1,5 @@
 #!/usr/bin/env bash
 
-set -euo pipefail
-
```
`set -euo pipefail` was messing up my interactive Bash session. It's already at the top of every normal script, and having it in these scripts that are meant to be imported with `source` messes up interactive Bash, so I removed it.

jamilbk reviewed (commented):
Just the note about a different approach for configuring packages

ReactorScram committed:
simplify the package list

ReactorScram committed:
ChatGPT recommended this

ReactorScram committed:
remove extra quotes

ReactorScram committed:
try again

ReactorScram committed:
typo

ReactorScram committed:
forgot to id this step

ReactorScram commented on .github/actions/setup-rust/action.yml:
```diff
@@ -1,3 +1,4 @@
+---
```
This isn't really necessary but it satisfies `yamllint` on my machine.

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into test/linux-ipc

Also minimized deps
